### PR TITLE
Add support for "week of month" in date parsing pattern

### PR DESCRIPTION
**Pull request by NEUpanning**
In date parsing pattern, "week of month" is used by `java.text.SimpleDateFormat`. So we need to support it for supporting SimpleDateFormat date parsing mode in DateTimeFormatter.  It's presented by number and the range is [1,5].

relates issue:#6374

### CODE REVIEW

**Comment by NEUpanning** [comment]
@PHILO-HE @mbasmanova Could you help to review this PR please? Thanks.

**Review by rui-mo — COMMENTED** [review]
Would you add some unit tests? Thanks.

File: velox/functions/lib/DateTimeFormatterBuilder.cpp [hunk 1]
```diff
@@ -56,6 +56,13 @@ DateTimeFormatterBuilder& DateTimeFormatterBuilder::appendWeekOfWeekYear(
   return *this;
 }
 
+DateTimeFormatterBuilder& DateTimeFormatterBuilder::appendWeekOfMonth(
```
Comment (PHILO-HE) [inline velox/functions/lib/DateTimeFormatterBuilder.cpp:59]: Should this function be called in buildJodaDateTimeFormatter or buildMysqlDateTimeFormatter? Like https://github.com/facebookincubator/velox/blob/main/velox/functions/lib/DateTimeFormatter.cpp#L1463.

Comment (NEUpanning) [reply velox/functions/lib/DateTimeFormatterBuilder.cpp:59]: No. "Week Of Month" is not supported in [Joda](https://www.joda.org/joda-time/key_format.html) and [Mysql](https://www.w3schools.com/sql/func_mysql_date_format.asp).

**Comment by NEUpanning** [comment]
@rui-mo Since "week of month" format is not supported in in [Joda](https://www.joda.org/joda-time/key_format.html) and [Mysql](https://www.w3schools.com/sql/func_mysql_date_format.asp) date patterns, I can't add unit tests using MysqlDateTimeFormatter or JodaDateTimeFormatter. Could I add unit tests in the PR supporting SimpleDateFormat laterly? Or do you have any other suggestions?

File: velox/type/TimestampConversion.h [hunk 2]
```diff
@@ -107,6 +107,15 @@ Status daysSinceEpochFromWeekDate(
     int32_t dayOfWeek,
     int64_t& out);
 
+/// Computes the (signed) number of days since unix epoch (1970-01-01).
+/// Returns UserError status if the date is invalid.
+Status daysSinceEpochFromWeekOfMonthDate(
```
Comment (mbasmanova) [inline velox/type/TimestampConversion.h:112]: Perhaps, extract this function into a separate PR and add a unit test.

File: velox/type/TimestampConversion.cpp [hunk 3]
```diff
@@ -593,6 +613,26 @@ Status daysSinceEpochFromWeekDate(
   return Status::OK();
 }
 
+Status daysSinceEpochFromWeekOfMonthDate(
+    int32_t year,
+    int32_t month,
+    int32_t weekOfMonth,
+    int32_t dayOfWeek,
+    int64_t& out) {
+  if (!isValidWeekOfMonthDate(year, month, weekOfMonth, dayOfWeek)) {
```
Comment (mbasmanova) [inline velox/type/TimestampConversion.cpp:622]: isValidWeekOfMonthDate doesn't perform full validation. For example, it allows weekOfMonth = 5 and dayOfWeek = 6, but these may not exist for all months. Where / how do we handle these cases of invalid inputs?

**Review by mbasmanova — COMMENTED** [review]
@NEUpanning As @rui-mo mentioned, it would be nice to add some tests. Some specific suggestions are below. Thanks.

**Comment by NEUpanning** [comment]
@mbasmanova @rui-mo I've added some unit tests in 7ec5bf492ddff8cd8f01992e0a22eb3818eb1457

File: velox/functions/lib/DateTimeFormatter.cpp [hunk 4]
```diff
@@ -735,7 +740,6 @@ int32_t parseFromPattern(
       return -1;
     }
     cur += size;
-    date.weekDateFormat = true;
```
Comment (rui-mo) [inline velox/functions/lib/DateTimeFormatter.cpp:738]: I wonder why we need this change, and could you clarify what is covered by `weekDateFormat` which seems to be a larger scope than `weekOfMonthDateFormat`?

**Review by rui-mo — COMMENTED** [review]
Thanks.

**Review by rui-mo — COMMENTED** [review]
@NEUpanning I notice some comments seems to be unresolved. Would you take a look first? Thanks.
https://github.com/facebookincubator/velox/pull/10511#pullrequestreview-2191598315 and https://github.com/facebookincubator/velox/pull/10511#pullrequestreview-2198174742

**Comment by NEUpanning** [comment]
@rui-mo The two comments you mentioned, one said "added test" and the other said "thanks". I have already added the test. Are there any other comments I missed?

**Comment by rui-mo** [comment]
> one said "added test" and the other said "thanks"

@NEUpanning Could you check the specific comments below them, e.g https://github.com/facebookincubator/velox/pull/10511#discussion_r1686566980? If they have been addressed, it would be nice to reply at the comment section.

File: velox/type/TimestampConversion.h (hunk 2 above)
Comment (rui-mo) [reply velox/type/TimestampConversion.h:112]: I assume this comment suggest us opening another PR which contains the implementation and tests for `daysSinceEpochFromWeekOfMonthDate` function.

**Comment by NEUpanning** [comment]
@rui-mo Sorry. I haven't submitted comments in a while. I think you can see it now.

File: velox/type/TimestampConversion.h (hunk 2 above)
Comment (NEUpanning) [reply velox/type/TimestampConversion.h:112]: I opened a PR for this. #10604

**Comment by NEUpanning** [comment]
Opened a new PR #11103 for this feature.
